### PR TITLE
Run AutoMod-blocked messages through autowarn filters

### LINKED ISSUE #209: Todo: Run autowarn filters on Automod-blocked messages
So say for example you have two lists, one in Cliptok and one in Discord Automod, your Automod list is designed to be less of a big deal than the autowarn list because it does not warn and simply prevents the message from being sent. If a user then sends a message matching both of these filters, the message will be blocked and not evaluated by the autowarn system. This is not ideal. The most severe punishment should be enacted regardless of whether the message was blocked or whether it was auto-deleted.

One way to solve this would be to integrate Cliptok into the Automod system entirely, by having it manage the lists and hook into Automod events to deliver warnings. Then no bad messages are sent at all, rejoice!

But that's too complicated, so we only need to do half of it for now: have Cliptok trigger on Automod-blocked messages as if they were the original message.

You would, in theory:
* Listen for Automod events.
  * Specifically message block events. Don't bother with message alerts (you will get the message event anyway) or username quarantines (these take care of themselves, no need for more punishment)
* From the event data, identify the user, channel and message content.
* Handle all the autowarn actions (i.e. ignore passive lists) in the Message Event that would normally trigger.
* If a warning or response is warranted, respond in the channel where the blocked message was attempted to be sent, and contact the user whose message was blocked.

The reason to avoid passive lists is to prevent duplicate alerts between what automod is posting in #investigations and what the passive list would post in the suspicious msgs feed. If we're not adding value (warning the user) then there is no need to take any action on the event.

I suspect the event is [AutoModerationRuleExecuted](https://dsharpplus.github.io/DSharpPlus/api/DSharpPlus.DiscordShardedClient.AutoModerationRuleExecuted.html) however I have not looked into how to use it. Amusingly even if that event did not work you could hack this together using the message event for the automod alert since it has a backwards-compatible Discord embed containing all the event data and a unique message type. I don't recommend it though.

### PR DESCRIPTION
**Pull request by FloatingMilkshake**
This PR closes #209. 

Runs messages that are blocked by AutoMod through our typical autowarn filters. This means users will be automatically warned for messages that would typically trip Cliptok's filters, even if AutoMod blocks the message from being sent first (currently this results in Cliptok not even seeing the message). This includes filters, like unapproved invites, mass emoji, scam checks, etc., but not 'passive' checks like user tracking*, passive list tracking, etc.

*Blocked messages will still end up in user-tracking threads in some way. It looks like this is just a blank embed pointing to the #investigations AutoMod message. This is because that message causes a Message Create event to be fired, and that is what makes it through to the tracking thread! I was initially not sure what to do about this, but I decided to leave it because I think it fits, considering how the point of the thread is to track all of the messages sent by a user.

AutoMod events don't give us a `DiscordMessage`, so this was a little awkward to do. I chose to create a fake `MockDiscordMessage` with the properties from the AutoMod event, and adjusted things like the message handler to accept that instead of `DiscordMessage` (there are now two overloads—one that takes a `MockDiscordMessage` and does the handling on that, and one that takes a `DiscordMessage`, throws its data into a `MockDiscordMessage`, and passes that to the other overload—this way there's no trouble when trying to use a regular `DiscordMessage`. I also did this for the `DeleteAndWarnAsync` and `MessageLink` methods). 

One thing I'm not sure about is the phrasing for when a user is pardoned the first time they send a message containing too many emoji or lines. For example, you can see here: https://github.com/Erisa/Cliptok/blob/dff0e2aede61fc5e3853ca32cea121632d0ab2c7/Events/MessageEvent.cs#L488-L491 ...and here: https://github.com/Erisa/Cliptok/blob/dff0e2aede61fc5e3853ca32cea121632d0ab2c7/Events/MessageEvent.cs#L623-L627 ...that the user sees a message like "Your message was deleted for mass emoji." I don't want to change that to "Your message was **blocked** ..." because the reason for the block might not be mass emoji—it could be something else, but Cliptok could be acting on mass emoji anyway. Open to suggestions on how this could be changed! :P

### CODE REVIEW

File: Helpers/InvestigationsHelpers.cs [hunk 1]
```diff
@@ -31,7 +35,10 @@ await LykosAvatarMethods.UserOrMemberAvatarURL(infringingMessage.Author, infring
                 embed.AddField(extraField.name, extraField.value, extraField.inline);
 
             if (content == default)
-                content = $"{Program.cfgjson.Emoji.Denied} Deleted infringing message by {infringingMessage.Author.Mention} in {infringingMessage.Channel.Mention}:";
+                if (wasAutoModBlock)
+                    content = $"{Program.cfgjson.Emoji.Denied} Detected infringing message (blocked by AutoMod) by {infringingMessage.Author.Mention} in {infringingMessage.Channel.Mention}:";
```
Comment (Erisa) [inline Helpers/InvestigationsHelpers.cs:39]: This feels too wordy
```suggestion
                    content = $"{Program.cfgjson.Emoji.Denied} Detected infringing AutoMod message by {infringingMessage.Author.Mention} in {infringingMessage.Channel.Mention}:";
```

**Review by Erisa — COMMENTED** [review]
Looks good, just one nitpick in the comments.

re: wording, I think it's best to be generic instead of situational with these.

"Play around with lots of emoji can stay as-is".

then:

```diff
-     msgOut = await message.Channel.SendMessageAsync($"{Program.cfgjson.Emoji.Information} {message.Author.Mention} Your message was automatically deleted for mass emoji."); 
+      msgOut = await message.Channel.SendMessageAsync($"{Program.cfgjson.Emoji.Information} {message.Author.Mention} Your message contained too many emoji."); 
```

```diff
-      string output = $"{Program.cfgjson.Emoji.Information} {message.Author.Mention}, your message was deleted for containing too many lines.\n" + 
+      string output = $"{Program.cfgjson.Emoji.Information} {message.Author.Mention}, your message contained too many lines.\n" + 
```

but also, can we please make the too many lines **not** offer the chance to show the full message, specifically for automod checks? It runs after other checks for a similar reason, which is to not act as an automod bypass opportunity.

File: Helpers/InvestigationsHelpers.cs (hunk 1 above)
Comment (FloatingMilkshake) [reply Helpers/InvestigationsHelpers.cs:39]: I agree but wasn't really sure how to put it. This works. Thanks!

**Comment by FloatingMilkshake** [comment]
> then:
> 
> ```diff
> -     msgOut = await message.Channel.SendMessageAsync($"{Program.cfgjson.Emoji.Information} {message.Author.Mention} Your message was automatically deleted for mass emoji."); 
> +      msgOut = await message.Channel.SendMessageAsync($"{Program.cfgjson.Emoji.Information} {message.Author.Mention} Your message contained too many emoji."); 
> ```
> 
> ```diff
> -      string output = $"{Program.cfgjson.Emoji.Information} {message.Author.Mention}, your message was deleted for containing too many lines.\n" + 
> +      string output = $"{Program.cfgjson.Emoji.Information} {message.Author.Mention}, your message contained too many lines.\n" + 
> ```
> 
> but also, can we please make the too many lines **not** offer the chance to show the full message, specifically for automod checks? It runs after other checks for a similar reason, which is to not act as an automod bypass opportunity.

Done! Good thinking with the automod bypass catch, that obviously didn't occur to me 😅

File: Events/MessageEvent.cs [hunk 2]
```diff
@@ -683,7 +685,9 @@ public static async Task MessageHandlerAsync(DiscordClient client, MockDiscordMe
                             string output = $"{Program.cfgjson.Emoji.Denied} {message.Author.Mention} was automatically warned: **{reason.Replace("`", "\\`").Replace("*", "\\*")}**\n" +
                                 $"Please consider using a Pastebin-style website or <#{Program.cfgjson.UnrestrictedEmojiChannels[0]}> to avoid punishment.";
                             DiscordMessageBuilder messageBuilder = new();
-                            messageBuilder.WithContent(output).AddComponents(button);
+                            messageBuilder.WithContent(output);
+                            if (!wasAutoModBlock)
+                                messageBuilder.AddComponents(button);
```
Comment (Erisa) [inline Events/MessageEvent.cs:690]: Lets avoid wasteful assignments and dead data by skipping the `var button = new DiscordButtonComponent` and `Program.db` interactions as well

**Review by Erisa — APPROVED** [review]
lgtm, merge when ready. The bug with some automod events not working correctly doesn't have any catastrophic consequences and doesn't appear to be the fault of this code.
